From dc4b2b76b3cc4cbd1dc8cc71e13d45caecb65d83 Mon Sep 17 00:00:00 2001
From: Hanjun Guo <guohanjun@huawei.com>
Date: Wed, 29 Sep 2021 16:02:06 +0800
Subject: [PATCH] arm64: topology: Support PHYTIUM CPU
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: dc4b2b76b3cc4cbd1dc8cc71e13d45caecb65d83
Modified-by-SEL: Yes, refreshed due to context change

hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4CGER
CVE: NA

---------------------------

Add the support for PHYTIUM topology detect, it's better
use PPTT ACPI table to report the topology, but we can live
with it at now.

Signed-off-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
Signed-off-by: Chenxi Mao <chenxi.mao@suse.com>
---
 arch/arm64/include/asm/cputype.h |    1 +
 arch/arm64/kernel/topology.c     |    5 +++++
 2 files changed, 6 insertions(+)

--- a/arch/arm64/include/asm/cputype.h
+++ b/arch/arm64/include/asm/cputype.h
@@ -60,6 +60,7 @@
 #define ARM_CPU_IMP_FUJITSU		0x46
 #define ARM_CPU_IMP_HISI		0x48
 #define ARM_CPU_IMP_APPLE		0x61
+#define ARM_CPU_IMP_PHYTIUM		0x70
 
 #define ARM_CPU_PART_AEM_V8		0xD0F
 #define ARM_CPU_PART_FOUNDATION		0xD00
--- a/arch/arm64/kernel/topology.c
+++ b/arch/arm64/kernel/topology.c
@@ -53,6 +53,11 @@ void store_cpu_topology(unsigned int cpu
 	cpuid_topo->thread_id  = -1;
 	cpuid_topo->core_id    = cpuid;
 	cpuid_topo->package_id = cpu_to_node(cpuid);
+	if (read_cpuid_implementor() == ARM_CPU_IMP_PHYTIUM) {
+		cpuid_topo->thread_id  = 0;
+		cpuid_topo->core_id  = cpuid;
+		cpuid_topo->package_id = 0;
+	}
 
 	pr_debug("CPU%u: cluster %d core %d thread %d mpidr %#016llx\n",
 		 cpuid, cpuid_topo->package_id, cpuid_topo->core_id,
